interface/ceed-qfunction.c

```diff
@@ -622,6 +628,54 @@ int CeedQFunctionAddOutput(CeedQFunction qf, const char *fieldname,
   return 0;
 }
 
+#ifdef CEED_CUDA_ENABLED
+/**
+  @brief Set CUDA function pointer to evaluate action at quadrature points
+
+  @param qf CeedQFunction to set device pointer
+  @param f  Device function pointer to evaluate action at quadrature points
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedQFunctionSetCUDAUserFunction(CeedQFunction qf, CUfunction f) {
+  int ierr;
+  if (!qf->SetCUDAUserFunction) {
+    Ceed ceed;
+    ierr = CeedQFunctionGetCeed(qf, &ceed); CeedChk(ierr);
+    CeedDebug("Backend does not support CUfunction pointers for QFunctions.");
+  } else {
+    ierr = qf->SetCUDAUserFunction(qf, f); CeedChk(ierr);
+  }
+  return 0;
+}
+#endif
+
+#ifdef CEED_HIP_ENABLED
+/**
+  @brief Set HIP function pointer to evaluate action at quadrature points
+
+  @param qf CeedQFunction to set device pointer
+  @param f  Device function pointer to evaluate action at quadrature points
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedQFunctionSetHIPUserFunction(CeedQFunction qf, hipFunction_t f) {
+  int ierr;
+  if (!qf->SetHIPUserFunction) {
+    Ceed ceed;
+    ierr = CeedQFunctionGetCeed(qf, &ceed); CeedChk(ierr);
+    CeedDebug("Backend does not support hipFunction_t pointers for QFunctions.");
+  } else {
+    ierr = qf->SetHIPUserFunction(qf, f); CeedChk(ierr);
+  }
+  return 0;
+}
+#endif
```

Comment:
Rather than having globally-defined macros `CEED_CUDA_ENABLED`, could we just move these functions to separate files that are only built whin CUDA (HIP, respectively) are available? That saves anti-patterns of checking those macros and avoids needing to force rebuild if adding/removing them from the configuration.

Make doesn't know when variables change, so if you
```console
make      # builds without CUDA
make CUDA_DIR=/path/to/cuda # builds CUDA sources
```
with the macros, this file won't be recompiled, leading to an inconsistent build. There are ways to ensure that make rebuilds in such conditions, but they are a blunt instrument and simpler is better when possible.

Reply:
I added a quick fix but didn't test. Headed to appointment, but can check later if CI issue.

Reply:
note to self, need to drop the defs of CEED_*_ENABLED

Reply:
Ok, rebased off main and fixed it up. It builds fine locally for CUDA and HIP